### PR TITLE
Use baseclasses in all builders. Rewrite tests

### PR DESCRIPTION
**Pull request by mallport**
This PR implements base classes that act as parent classes for the builders `Pseudonymize`, `Depseudonymize` and `Repseudonymize`. These base classes contain common functionality that is shared across all builders, hopefully making maintenance easier. 

The PR also refactors the tests. More is explained below.

There should be no functional changes in this PR.

# Changes 

## Baseclasses

There are two baseclasses, containing the shared logic used by all builders. These baseclasses replace, as an example for the `Pseudonymize` builder, the `_Pseudonymizer` and `_PseudoFuncSelector` classes, and can be found in [this file](https://github.com/statisticsnorway/dapla-toolbelt-pseudo/blob/c9ef77ba3149d830b9f059cdb9943765ebd26c75/src/dapla_pseudo/v1/baseclasses.py).

While it is not necessary to use inheritance and base classes, one _could_ plausibly use functions instead, the abstraction of having a constructor in the base class is useful. The constructor of the base class displays the state that is known when the child builder is first created, and the parameters in the base class methods represent the state that is passed through the user-facing methods.

## Tests

The tests used previously was a patchwork of differing philosophies. The tests were becoming increasingly less ergonomic and rife with mocks that are hard to understand due to their dynamic nature.

The fact that `pytest` puts severe restrictions on parametrization only worsened this issue. You cannot parametrize over objects, fixtures or types - only primitive types such as string or int are allowed. 

This means that tests need to be either be:

1. duplicated for every case 
2. supply all the primitives needed to create an object in the parametrization

As the code base is quite object-oriented by nature, option 2) would lead to a huge number of parametrized arguments in order to construct the nested objects, hurting readability. Thus, the easiest way to write tests was usually to duplicate for every case you wanted to cover. This lead to a much higher maintenance burden over time.

**However**, this PR rewrites the tests to use the library [pytest-cases](https://smarie.github.io/python-pytest-cases/), which changes the underlying `pytest`-engine, opening up to parametrize over almost all types. This allows much more modular tests to be written.

Tests are written for the base classes, but _not_ for the child classes, i.e. the user-exposed builders. The logic inside the builders is now so rudimentary that the value of having tests for the builders is questionable. In my opinion, spending energy on writing integration tests exposes errors at a much lower maintenance cost.

## Refactor

Previously, we only had one file for the `pydantic` models, `api_models.py`. However, how they were used cut quite cleanly along two categories. One category of models were the API-facing models, representing requests and responses from the API, while the other category represented pseudonymization building blocks (e.g. PseudoFunction).

The models have been split into two files:
```
.
└── models/
    ├── api.py
    └── core.py
```

`pseudo_commons.py` has also been deleted, much of the functionality has been absorbed into the baseclasses or `utils.py`. Arguably, the functions creating the requests in `utils.py` don't belong there, I'm open to suggestions as to where they could live.

### CODE REVIEW

File: src/dapla_pseudo/v1/baseclasses.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+                response = self._pseudo_client._post_to_file_endpoint(
+                    path=f"{self._pseudo_operation.value}/file",
+                    request_spec=request_spec,
+                    data_spec=data_spec,
+                    stream=True,
+                    timeout=timeout,
+                )
+                file_handle.close()
+            case HierarchicalDataFrame(dataset):
```
Comment (bjornandre) [inline src/dapla_pseudo/v1/baseclasses.py:203]: HierarchicalDataFrame was removed in #383

File: src/dapla_pseudo/v1/baseclasses.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            case File():
+                pseudo_request = build_pseudo_dataset_request(
+                    self._pseudo_operation,
+                    rules,
+                    custom_keyset,
+                    target_custom_keyset,
+                    target_rules,
+                )
+                pseudo_response = self._pseudonymize_dataset(pseudo_request, timeout)
+            case HierarchicalDataFrame():
```
Comment (bjornandre) [inline src/dapla_pseudo/v1/baseclasses.py:102]: HierarchicalDataFrame was removed in #383

File: src/dapla_pseudo/v1/baseclasses.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        )
+
+
+class _BaseRuleConstructor:
+    """Base class for the _PseudoFuncSelector/_DepseudoFuncSelector/_RepseudoFuncSelector builders."""
+
+    def __init__(
+        self,
+        fields: list[str],
+        dataset_type: type[pl.DataFrame] | type[File] | type[HierarchicalDataFrame],
```
Comment (bjornandre) [inline src/dapla_pseudo/v1/baseclasses.py:242]: HierarchicalDataFrame was removed in #383

**Review by bjornandre — CHANGES_REQUESTED** [review]

File: src/dapla_pseudo/v1/baseclasses.py (hunk 3 above)
Comment (mallport) [reply src/dapla_pseudo/v1/baseclasses.py:242]: Aha, I didn't see this PR :) Merge went a bit quickly I see

File: src/dapla_pseudo/__init__.py [hunk 4]
```diff
@@ -20,11 +20,9 @@
 # Ref: https://github.com/python-poetry/poetry/issues/144#issuecomment-1488038660
 __version__ = version("dapla_toolbelt_pseudo")
 
-from dapla_pseudo.v1.api_models import Field
-from dapla_pseudo.v1.api_models import PseudoKeyset
-from dapla_pseudo.v1.api_models import PseudoRule
```
Comment (bjornandre) [inline src/dapla_pseudo/__init__.py:25]: Can you add this again please? I think `PseudoRule` should be easily accessible, ref this use-case: https://github.com/statisticsnorway/dapla-toolbelt-pseudo?tab=readme-ov-file#pseudonymize-using-custom-rules

**Review by bjornandre — CHANGES_REQUESTED** [review]

File: src/dapla_pseudo/__init__.py (hunk 4 above)
Comment (mallport) [reply src/dapla_pseudo/__init__.py:25]: Oops, I didn't intend to remove that. Added back

**Review by bjornandre — APPROVED** [review]

**Pull request merged**